### PR TITLE
add relative_to for formatter_for_file

### PR DESCRIPTION
**Pull request by DianaOlympos**
This address https://github.com/elixir-lsp/elixir-ls/issues/526 and https://groups.google.com/g/elixir-lang-core/c/r5FA-4_Fu9o/m/rZiFgGOJBAAJ

In the case of IDE trying to format snippets or files, it is possible that another task is running concurrently that may change `cwd` while we are recovering the formatter options. This end up showing as the formatter failing. It is particularly visible for people using "Format on Save", as recompilation usually happens at save time.

To avoid the problem, I introduce a new option as discussed above, `:relative_to`. I did not expand the general `mix format` to accept the option, as it is right now only documented to work from the root of the project and I did not see a need for it.

I reused the `prefix` concept that was already existing for plugins, apps, deps, and subdirectories, as I considered that it was quite close to what I am doing here. I added a simple test, I do not know if you want more. I am also not a fan of the name of the option, maybe `:root_path` would make more sense?

### CODE REVIEW

**Comment by DianaOlympos** [comment]
My only fear with `cwd` is if someone mix it with the output of `File.cwd` but you know the conventions of the code base better than me.


On the filename logic. I think there are two things here. The first is the input list, for plugins and all. The second is that when we go for subdirectories and deps recursively, we need to handle the prefix when recurring. I am not sure how important it is but it is what i see in the code i changed? I mean first time i looked into it was on Sunday :D

File: lib/mix/test/mix/tasks/format_test.exs [hunk 1]
```diff
@@ -523,6 +523,28 @@ defmodule Mix.Tasks.FormatTest do
...
+  test "uses inputs and configuration from :relative_to path", context do
+    in_tmp(context.test, fn ->
+      File.write!(".formatter.exs", """
+      [
+        locals_without_parens: [foo: 1]
+      ]
+      """)
+
+      root = File.cwd!()
+      File.cd("lib")
```
Comment (josevalim) [inline lib/mix/test/mix/tasks/format_test.exs:535]: You need to revert that so it doesn't affect other tests.

**Comment by josevalim** [comment]
Regarding `relative_to`, I was wondering if instead of doing `relative_to`, we could change it so `Path.expand/1` receives the `cwd` as second argument, so the expansion is correct?

https://github.com/elixir-lang/elixir/blob/419f6aa424e634b83f67575f471e71b1af8395e0/lib/mix/lib/mix/tasks/format.ex#L612

**Comment by DianaOlympos** [comment]
I am not sure if that would work for plugins and subdirectories though. Will have to think through the cases tomorrow.

File: lib/mix/test/mix/tasks/format_test.exs (hunk 1 above)
Comment (whatyouhide) [reply lib/mix/test/mix/tasks/format_test.exs:535]: You can use `File.cd!/2` 🙃

Comment (DianaOlympos) [reply lib/mix/test/mix/tasks/format_test.exs]: i had totally forgotten this one @whatyouhide thanks!

**Comment by DianaOlympos** [comment]
@josevalim After looking into it, we could do that, but this function is only used by `run`, not by `formatter_for_file/2` or `formatter_opts_for_file/2`. So it would not solve the problem for the LSP.

I don't know how much work it would be if we refactor the whole thing to use this whole path filtering. This is not only a Mix.Task. This is also the logic to generate the formatting function, based on the options, that is a public API.

What do you prefer? `cwd` or `root` or `relative_to`?

**Comment by DianaOlympos** [comment]
I will update that name tomorrow morning.

**Comment by DianaOlympos** [comment]
Should be good now!

**Comment by josevalim** [comment]
@DianaOlympos tests are failing. Can you please take a look? :)

**Comment by DianaOlympos** [comment]
ok should be fixed now. No idea why these disappeared, I was pretty sure I had put them in. Well anyway, not hard to fix.

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

File: lib/mix/lib/mix/tasks/format.ex [hunk 2]
```diff
@@ -251,10 +251,12 @@ defmodule Mix.Tasks.Format do
   def formatter_for_file(file, opts \\ []) do
     {dot_formatter, formatter_opts} = eval_dot_formatter(opts)
 
+    prefix = Keyword.get(opts, :root, [])
+
     {formatter_opts_and_subs, _sources} =
-      eval_deps_and_subdirectories(dot_formatter, [], formatter_opts, [dot_formatter])
+      eval_deps_and_subdirectories(dot_formatter, prefix, formatter_opts, [dot_formatter])
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/format.ex:257]: This is broken. `prefix` is meant to be a list and if you pass `:root` as a string, it won't work.

Comment (josevalim) [reply lib/mix/lib/mix/tasks/format.ex:257]: I am on it.

**Comment by josevalim** [comment]
Sorry, I have to revert this PR. There is still a lot of relative_to_cwd and cwd checks and I am not confident those won't be broken. I think the best option forward is to compute `cwd` at the top-level and pass it as argument throughout all functions.

**Comment by DianaOlympos** [comment]
Oh urgh i am sorry for this ended up being like that. I was really not confident on the tests but I felt like the existing edge cases were already not handled... Sorry :(

**Comment by josevalim** [comment]
No worries :)
